### ISSUE TITLE
traceconv get empty file

### ISSUE THREAD

**Issue by tingfengwanxiao** [google/perfetto#951]
When I use as ./traceconv.exe systrace 1.pftrace 2.ftrace,I get a empty 2.ftrace,
it seems because my 1.pftrace is too large as 8G,so How can I change so large perfetto trace to ftrace,
or how can I directly capture a ftrace by som command as large as 8G?

**Comment by LalitMaganti** [comment]
Are you sure it's actually because the file is large? Have you tried with a smaller file and things work correctly?

I would be a bit surprised if the file being large was an issue albeit it's not impossible I guess. 

Just to be clear, if you have an 8GB Perfetto file, that will turn into a 20GB or more systrace file. What are you planning on doing with that in the end?

**Comment by tingfengwanxiao** [comment]
I want to capture a long-term trace and then divide it into smaller ones at certain time points, but perfetto trace seems unable to be directly divided, so I want to convert it into ftrace and use text processing to divide it. Do you have any good suggestions

**Comment by LalitMaganti** [comment]
Why not just collect many small traces in the first place? Would be a lot easier than capturing a big trace and then trying to split it.

**Comment by LalitMaganti** [comment]
Fine I don't think it's worth discussing more in that direction.

Back to the actual topic at hand: would still be useful to know if this is only an issue on big traces or also happening on smaller traces. Especially interesting is if the cut-off is 4GB.

**Comment by LalitMaganti** [comment]
Do you have enough memory on your system to have the full trace in memory i.e. at least 20GB of free memory? Because systrace conversion works by first parsing the full trace into trace processor tables before then doing the conversion.

**Comment by LalitMaganti** [comment]
Specifically, what happens if you just try and just parse the trace with trace_processor_shell? Does it work?

**Comment by tingfengwanxiao** [comment]
> Do you have enough memory on your system to have the full trace in memory i.e. at least 20GB of free memory? Because systrace conversion works by first parsing the full trace into trace processor tables before then doing the conversion.

I only have 16G,maybe its reason

**Comment by tingfengwanxiao** [comment]
Then how can I do for this large trace? split it  by time for some small trace

**Comment by LalitMaganti** [comment]
There's not really a good solution for this today unfortunately. Everyone who needs this sort of thing today just collects smaller traces in the first place.

**Comment by tingfengwanxiao** [comment]
ok ,thank you !

**Comment by LalitMaganti** [comment]
I've changed this bug to track a better implementation of trace splitting: we might implement it at some point but for us it's not a super high priority.

**Comment by tingfengwanxiao** [comment]
Or whether it is possible to read in batches instead of all at once

> Do you have enough memory on your system to have the full trace in memory i.e. at least 20GB of free memory? Because systrace conversion works by first parsing the full trace into trace processor tables before then doing the conversion.

**Comment by LalitMaganti** [comment]
Also to use trace processor with such a large trace in the first place still requires you to have enough memory to hold the full trace in memory.

**Comment by ManiacLook** [comment]
> Why not just collect many small traces in the first place? Would be a lot easier than capturing a big trace and then trying to split it.

I used the parameters `max_file_size_bytes` and `duration_ms` shown in the following images, but could only obtain one trace file.
For example, if I set the `max_file_size_bytes` to 10000, then when I obtain a trace data volume of 10000, it will not continue to generate trace files.

I want to automatically collect multiple small trace files that are continuous and seamless at different times when collecting trace data. Is there a way to achieve this?

![Image](https://github.com/user-attachments/assets/17d428df-ee01-4212-b646-a872024ef5a0)

![Image](https://github.com/user-attachments/assets/2f2d4c4c-1e35-4049-a340-4d12d0b16e48)

**Comment by fengzikuaipao** [comment]
> Why not just collect many small traces in the first place? Would be a lot easier than capturing a big trace and then trying to split it.

Hi，I want to obtain a bunch of trace files with long time spans, but there is no time gap between the files. But the trace config will stop the trace when the limit was reached. When I restart a new trace, the time gap will occur.

![Image](https://github.com/user-attachments/assets/10dc5ce8-0e8a-44c7-92ab-d24845b91c2a)

![Image](https://github.com/user-attachments/assets/04d79726-debd-4e31-8804-54588230d27a)

Do we have the function of automatic slicing during collection?

**Comment by LalitMaganti** [comment]
Copying from https://github.com/google/perfetto/pull/2020 for future readers who want to know about why we don't support trace splitting:

> It's a very hard problem to do in a general way, you're going to run in all sorts of subtle issue with "ambient state" (things like which processes are running on the system) if you split traces naively.

> Concretely: at the start of traces, we snapshot the full state of all the processes on the system. If you do a naive splitting tool where you just split the trace based on packet boundaries, you'll drop this packet and then be confused why there in your split trace all the processes are missing.

> Now there are probably 20-25 different similar instances of "we snapshot this at the start of the trace and we want to preserve it" and many more of "we want to preserve some data (e.g. the value of a counter) from before the start of the split trace".

> So TLDR it's a very hard problem. It'll be something we get to at some point but to do it properly, it's a multi-month project and it's never been a priority so not something we will get to soon.

**Comment by douglas-raillard-arm** [comment]
@LalitMaganti I'm looking into splitting traces as well in our conversion pipeline so that we could process one batch at a time and reduce peak memory consumption. So far, everything downstream of Perfetto is chunk-oriented: we issue multiple "paged" SQL queries (with `WHERE ftrace_event.id > x LIMIT n`) and convert to pandas then dump in parquet. The only missing part is having `trace_processor` ingest a partial trace in the first place.

This pipeline only uses the raw ftrace events data, so we don't really need any state to be preserved.

In that context, assuming I had some generic protobuf "scissors" and just cut the stream at a packet boundary, can I expect `trace_processor` to be able to ingest that ? Or will the absence of some "state packets" break some things internally beyond e.g. empty `thread` table and make it error/crash ?

**Comment by LalitMaganti** [comment]
> In that context, assuming I had some generic protobuf "scissors" and just cut the stream at a packet boundary, can I expect trace_processor to be able to ingest that ?

So I **think** it should work but I don't think we can guarantee this will stay the case into the future. We are having active discussions in how the core tracing protocol and service will change. That might require knock on impacts on how the protos are structured (i.e. "columnarizing" the protos to reduce trace size) and there might also be dependencies on interning etc.

**Comment by douglas-raillard-arm** [comment]
Right, I'll hold off until this happens and reconsider then. Thanks.

**Comment by WWKy** [comment]
@LalitMaganti Hi~, 
We are using the Perfetto SDK on Android and need to implement time-based trimming of trace data on the device.

**Scenario:**
We initiate a Perfetto session at system boot that streams data into a ring buffer. When specific events occur, we stop the session and use `ReadTrace(ReadTraceCallback)` to read trace packets in batches. To save space, we need to dynamically trim the trace to 3s, 5s, or keep it full based on the event type before saving.

**Current Progress & Issues:**
When using only `linux.ftrace`, we tested tokenizing data in the ReadTraceCallback to check `ftrace_event` timestamps. If a timestamp is within the target range, we keep the batch; otherwise, we discard it. The resulting merged file works fine in the Perfetto UI.

However, after adding `linux.perf`, trimming has become difficult because:
1. `linux.perf` and `linux.ftrace` use different default clock sources.
2. The trace packets from these two sources are not globally ordered across batches.

Does the Perfetto team have any recommendations for this requirement?

**Comment by LalitMaganti** [comment]
Yeah sorry this is *exactly* why this problem is so hard to solve. There's no easy general purpose solution to this without reinventing tons of code which lives in trace processor today (e.g. clock sync, which is probably one of the most subtle classes in TP). 

The situation remains that this is a very hard problem which it would be good to solve but the benefits of this are unclear. Unless there's lots of traction on this bug, it'll be an aspirational goal but not one which will solve immediately. At the very least, I would like to solve *merging* traces before splitting as there's a lot more interesting things you can do with trace merging (i.e. you can collect traces separately and merge post facto).
